File: apps/gateway/src/agent/tools/run-command.ts
// run_command 核心：在绑定 pane 里执行一条命令并拿到完整输出。
// 数据全部走实时流（emulator tap）：字节累积命令输出、OSC133 标记判完成、render 判 alternate。
// 三类目标（详见 docs）：
//  - POSIX：注入隐形 OSC133 + nonce 包裹命令，等带 nonce 的 D 标记 → 精确输出 + 退出码。
//  - CLI（网络设备）：学提示符 / 提示符重现判完成 / --More-- 自动续翻 / 错误串启发（无退出码）。
//  - TUI/alternate：拒绝（entered_tui），交回交互式读写屏。

import type { PromptMarker } from '../../tmux-client/pane-stream-parser';

export type RunCommandMode = 'auto' | 'posix' | 'cli';
export type RunCommandShell = 'bash' | 'zsh' | 'sh' | 'fish' | 'powershell';

export type RunCommandStatus =
  | 'completed'
  | 'timeout'
  | 'entered_tui'
  | 'expect_matched'
  | 'paused_pager';

export interface RunCommandResult {
  output: string;
  exitCode: number | null;
  status: RunCommandStatus;
  likelyError: boolean;
  errorLine?: string;
  truncated: boolean;
}

/** run_command 所需的 emulator 能力子集（便于以 fake 单测）。 */
export interface RunCommandEmulator {
  isAlternateScreen(): boolean;
  render(): string;
  tap(tap: { onBytes?: (data: Uint8Array) => void; onMarker?: (marker: PromptMarker) => void }): () => void;
}

export interface RunCommandParams {
  command: string;
  mode?: RunCommandMode;
  shell?: RunCommandShell;
  /** cli 提示符正则（不传则从当前屏末行学习） */
  prompt?: string;
  /** 命中即早返回（密码提示 / [y/N] 等） */
  expect?: string;
  timeoutMs?: number;
  /** cli：先发该平台关分页命令 */
  disablePagingCommand?: string;
}

export interface RunCommandDeps {
  emulator: RunCommandEmulator;
  sendInput: (data: string) => void;
  sleepMs?: (ms: number) => Promise<void>;
  /** 注入 nonce（默认基于计数器；避免直接用 Math.random，便于测试） */
  makeNonce?: () => string;
  now?: () => number;
}

const DEFAULT_TIMEOUT_MS = 15_000;
const OUTPUT_MAX_BYTES = 256 * 1024;
const POLL_MS = 50;
const MORE_MARKERS = [/--More--/, /---\(more[^)]*\)---/i, /<--- More --->/i, /\bMore: <space>/i];
const ERROR_PATTERNS = [
  /%\s*Invalid input/i,
  /%\s*Ambiguous command/i,
  /%\s*Incomplete command/i,
  /^\s*\^\s*$/m,
  /syntax error/i,
  /unknown command/i,
];

const decoder = new TextDecoder();

function exitCodeExpr(shell: RunCommandShell | undefined): string | null {
  switch (shell) {
    case 'fish':
      return '$status';
    case 'powershell':
      return null; // pwsh 转义不同，退回无退出码路径
    default:
      return '$?'; // bash/zsh/sh 及未知
  }
}

// 去 ANSI/控制序列 + 处理 \r 覆盖 + 整理空白，得到干净文本。
export function cleanTerminalText(raw: string): string {
  let text = raw
    // OSC: ESC ] ... (BEL | ESC \)
    .replace(/\x1b\][^\x07\x1b]*(?:\x07|\x1b\\)/g, '')
    // CSI: ESC [ ... letter
    .replace(/\x1b\[[0-9;?]*[ -/]*[@-~]/g, '')
    // 其它两字节/字符集切换转义
    .replace(/\x1b[()][AB0-9]/g, '')
    .replace(/\x1b[=>NOcDEHM]/g, '')
    // 退格
    .replace(/.\x08/g, '');
  // 处理 \r：行内回车覆盖，保留最后一段
  text = text
    .split('\n')
    .map((line) => {
      if (!line.includes('\r')) {
        return line.replace(/\s+$/, '');
      }
      const segs = line.split('\r').filter((s) => s.length > 0);
      return (segs[segs.length - 1] ?? '').replace(/\s+$/, '');
    })
    .join('\n');
  return text;
}

function lastNonEmptyLine(text: string): string {
  const lines = cleanTerminalText(text).split('\n');
  for (let i = lines.length - 1; i >= 0; i -= 1) {
    if (lines[i].trim().length > 0) {
      return lines[i];
    }
  }
  return '';
}

// 从累积原始字节里剥掉第一行（命令回显），返回干净输出。
function extractOutput(raw: string): { text: string; truncated: boolean } {
  const truncated = raw.length >= OUTPUT_MAX_BYTES;
  const cleaned = cleanTerminalText(raw);
  const newlineIdx = cleaned.indexOf('\n');
  // 第一行是 shell 对输入行的回显（含我们注入的 wrapper），剥掉
  const body = newlineIdx >= 0 ? cleaned.slice(newlineIdx + 1) : '';
  return { text: body.replace(/\n+$/, ''), truncated };
}

function detectError(text: string): { likelyError: boolean; errorLine?: string } {
  for (const pattern of ERROR_PATTERNS) {
    const match = pattern.exec(text);
    if (match) {
      const line = text.split('\n').find((l) => pattern.test(l)) ?? match[0];
      return { likelyError: true, errorLine: line.trim() };
    }
  }
  return { likelyError: false };
}

export async function executeRunCommand(
  params: RunCommandParams,
  deps: RunCommandDeps
): Promise<RunCommandResult> {
  const sleepMs = deps.sleepMs ?? ((ms: number) => new Promise((r) => setTimeout(r, ms)));
  const now = deps.now ?? (() => performance.now());
  let nonceCounter = 0;
  const makeNonce = deps.makeNonce ?? (() => `n${++nonceCounter}${(now() | 0).toString(36)}`);
  const timeoutMs = params.timeoutMs ?? DEFAULT_TIMEOUT_MS;
  const mode: RunCommandMode = params.mode ?? 'auto';

  if (deps.emulator.isAlternateScreen()) {
    return {
      output: '',
      exitCode: null,
      status: 'entered_tui',
      likelyError: false,
      truncated: false,
    };
  }

  // 累积流字节 + 标记
  const chunks: number[] = [];
  let receivedMarker: PromptMarker | null = null;
  let nonce = '';
  const untap = deps.emulator.tap({
    onBytes: (data) => {
      for (const byte of data) {
        if (chunks.length < OUTPUT_MAX_BYTES) {
          chunks.push(byte);
        }
      }
    },
    onMarker: (marker) => {
      if (marker.kind === 'D' && (!nonce || marker.params.includes(`tmex=${nonce}`))) {
        receivedMarker = marker;
      }
    },
  });

  const accumulated = (): string => decoder.decode(new Uint8Array(chunks));
  const finish = (result: RunCommandResult): RunCommandResult => {
    untap();
    return result;
  };

  try {
    const usePosix =
      mode === 'posix' || (mode === 'auto' && exitCodeExpr(params.shell) !== null);

    // cli：可选先关分页
    if (mode === 'cli' && params.disablePagingCommand) {
      deps.sendInput(`${params.disablePagingCommand}\r`);
      await sleepMs(200);
      chunks.length = 0;
    }

    // 学习 cli 提示符（用于完成判定）
    const promptRegex = params.prompt
      ? new RegExp(params.prompt)
      : mode === 'cli'
        ? buildPromptRegex(lastNonEmptyLine(deps.emulator.render()))
        : null;

    // 发送命令（POSIX 包裹隐形 OSC133 + nonce）
    if (usePosix) {
      nonce = makeNonce();
      const expr = exitCodeExpr(params.shell) ?? '$?';
      const marker = `printf '\\033]133;D;%s;tmex=${nonce}\\033\\\\' "${expr}"`;
      deps.sendInput(`${params.command}; ${marker}\r`);
    } else {
      deps.sendInput(`${params.command}\r`);
    }

    const expectRegex = params.expect ? new RegExp(params.expect) : null;
    const deadline = now() + timeoutMs;
    let idleStableSince = 0;
    let lastLen = 0;

    while (now() < deadline) {
      await sleepMs(POLL_MS);

      // alternate 屏（命令切进 TUI，如 vim/less）
      if (deps.emulator.isAlternateScreen()) {
        return finish({
          output: extractOutput(accumulated()).text,
          exitCode: null,
          status: 'entered_tui',
          likelyError: false,
          truncated: false,
        });
      }

      const rawNow = accumulated();
      const cleanedNow = cleanTerminalText(rawNow);

      // expect 命中
      if (expectRegex?.test(cleanedNow)) {
        const out = extractOutput(rawNow);
        return finish({
          output: out.text,
          exitCode: null,
          status: 'expect_matched',
          likelyError: false,
          truncated: out.truncated,
        });
      }

      // POSIX：等到带 nonce 的 D 标记
      if (usePosix && receivedMarker) {
        const out = extractOutput(rawNow);
        const err = detectError(out.text);
        return finish({
          output: out.text,
          exitCode: (receivedMarker as PromptMarker).exitCode,
          status: 'completed',
          ...err,
          truncated: out.truncated,
        });
      }

      // 分页：遇 --More-- 自动续翻
      if (MORE_MARKERS.some((re) => re.test(cleanedNow.slice(-200)))) {
        deps.sendInput(' ');
        idleStableSince = 0;
        continue;
      }

      // CLI / 非 POSIX：提示符在末尾重现 = 完成
      if (promptRegex) {
        const tail = lastNonEmptyLine(rawNow);
        const out = extractOutput(rawNow);
        if (promptRegex.test(tail) && out.text.length > 0) {
          const err = detectError(out.text);
          return finish({
            output: out.text,
            exitCode: null,
            status: 'completed',
            ...err,
            truncated: out.truncated,
          });
        }
      }

      // 输出静默判定（auto/posix 无标记回退）
      if (rawNow.length === lastLen) {
        if (idleStableSince === 0) {
          idleStableSince = now();
        } else if (now() - idleStableSince >= 600) {
          const out = extractOutput(rawNow);
          if (out.text.length > 0 || now() - idleStableSince >= 1500) {
            const err = detectError(out.text);
            return finish({
              output: out.text,
              exitCode: null,
              status: 'completed',
              ...err,
              truncated: out.truncated,
            });
          }
        }
      } else {
        idleStableSince = 0;
        lastLen = rawNow.length;
      }
    }

    // 超时
    const out = extractOutput(accumulated());
    return finish({
      output: out.text,
      exitCode: null,
      status: 'timeout',
      ...detectError(out.text),
      truncated: out.truncated,
    });
  } finally {
    untap();
  }
}

function buildPromptRegex(promptLine: string): RegExp | null {
  const trimmed = promptLine.trim();
  if (!trimmed) {
    return null;
  }
  const escaped = trimmed.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
  return new RegExp(`${escaped}\\s*$`);
}

File: apps/gateway/src-rust/agent/tools/run_command.rs
use std::time::Duration;

use async_trait::async_trait;
use regress::Regex;
use tokio::time::Instant;

use crate::agent::AgentPortError;

pub const RUN_COMMAND_DEFAULT_TIMEOUT: Duration = Duration::from_secs(15);
pub const RUN_COMMAND_OUTPUT_MAX_BYTES: usize = 256 * 1024;
pub const RUN_COMMAND_POLL_INTERVAL: Duration = Duration::from_millis(50);

#[derive(Clone, Copy, Debug, Default, PartialEq, Eq)]
pub enum RunCommandMode {
    #[default]
    Auto,
    Posix,
    Cli,
}

#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum RunCommandShell {
    Bash,
    Zsh,
    Sh,
    Fish,
    PowerShell,
}

#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum RunCommandStatus {
    Completed,
    Timeout,
    EnteredTui,
    ExpectMatched,
    PausedPager,
}

#[derive(Clone, Debug, PartialEq, Eq)]
pub struct RunCommandResult {
    pub output: String,
    pub exit_code: Option<i32>,
    pub status: RunCommandStatus,
    pub likely_error: bool,
    pub error_line: Option<String>,
    pub truncated: bool,
}

#[derive(Clone, Debug, PartialEq, Eq)]
pub struct RunCommandParams {
    pub command: String,
    pub mode: RunCommandMode,
    pub shell: Option<RunCommandShell>,
    pub prompt: Option<String>,
    pub expect: Option<String>,
    pub timeout: Duration,
    pub disable_paging_command: Option<String>,
}

impl RunCommandParams {
    pub fn new(command: impl Into<String>) -> Self {
        Self {
            command: command.into(),
            mode: RunCommandMode::Auto,
            shell: None,
            prompt: None,
            expect: None,
            timeout: RUN_COMMAND_DEFAULT_TIMEOUT,
            disable_paging_command: None,
        }
    }
}

#[derive(Clone, Debug, PartialEq, Eq)]
pub struct CommandPromptMarker {
    pub kind: char,
    pub exit_code: Option<i32>,
    pub params: Vec<String>,
}

#[derive(Clone, Debug, Default, PartialEq, Eq)]
pub struct CommandStreamUpdate {
    pub bytes: Vec<u8>,
    pub markers: Vec<CommandPromptMarker>,
    pub alternate_screen: bool,
}

#[async_trait]
pub trait TerminalCommandStream: Send {
    fn initial_screen(&self) -> &str;
    fn initial_alternate_screen(&self) -> bool;
    async fn send_input(&mut self, data: &str) -> Result<(), AgentPortError>;
    async fn poll(&mut self, wait: Duration) -> Result<CommandStreamUpdate, AgentPortError>;
    async fn close(&mut self);
}

async fn discard_stream_updates_for(
    stream: &mut dyn TerminalCommandStream,
    duration: Duration,
) -> Result<(), AgentPortError> {
    let deadline = Instant::now() + duration;
    loop {
        let now = Instant::now();
        if now >= deadline {
            return Ok(());
        }
        let _ = stream.poll(deadline.saturating_duration_since(now)).await?;
    }
}

fn exit_code_expression(shell: Option<RunCommandShell>) -> Option<&'static str> {
    match shell {
        Some(RunCommandShell::Fish) => Some("$status"),
        Some(RunCommandShell::PowerShell) => None,
        _ => Some("$?"),
    }
}

pub fn clean_terminal_text(raw: &str) -> String {
    let bytes = raw.as_bytes();
    let mut clean = Vec::with_capacity(bytes.len());
    let mut index = 0;
    while index < bytes.len() {
        if bytes[index] != 0x1b {
            if bytes[index] == 0x08 {
                clean.pop();
            } else {
                clean.push(bytes[index]);
            }
            index += 1;
            continue;
        }

        index += 1;
        let Some(next) = bytes.get(index).copied() else {
            break;
        };
        match next {
            b']' => {
                index += 1;
                while index < bytes.len() {
                    if bytes[index] == 0x07 {
                        index += 1;
                        break;
                    }
                    if bytes[index] == 0x1b && bytes.get(index + 1) == Some(&b'\\') {
                        index += 2;
                        break;
                    }
                    index += 1;
                }
            }
            b'[' => {
                index += 1;
                while index < bytes.len() {
                    let byte = bytes[index];
                    index += 1;
                    if (0x40..=0x7e).contains(&byte) {
                        break;
                    }
                }
            }
            b'(' | b')' => index = (index + 2).min(bytes.len()),
            _ => index += 1,
        }
    }

    String::from_utf8_lossy(&clean)
        .split('\n')
        .map(|line| {
            line.rsplit('\r')
                .find(|segment| !segment.is_empty())
                .unwrap_or_default()
                .trim_end()
        })
        .collect::<Vec<_>>()
        .join("\n")
}

fn last_non_empty_line(text: &str) -> &str {
    text.lines()
        .rev()
        .find(|line| !line.trim().is_empty())
        .unwrap_or_default()
}

fn prompt_pattern(prompt_line: &str) -> Option<Regex> {
    let trimmed = prompt_line.trim();
    if trimmed.is_empty() {
        return None;
    }
    let escaped = trimmed
        .chars()
        .flat_map(|character| {
            if ".*+?^${}()|[]\\".contains(character) {
                vec!['\\', character]
            } else {
                vec![character]
            }
        })
        .collect::<String>();
    Regex::new(&format!(r"{escaped}\s*$")).ok()
}

fn extract_output(raw: &[u8], overflowed: bool) -> (String, bool) {
    let cleaned = clean_terminal_text(&String::from_utf8_lossy(raw));
    let body = cleaned
        .split_once('\n')
        .map_or("", |(_, body)| body)
        .trim_end_matches('\n')
        .to_owned();
    (
        body,
        overflowed || raw.len() >= RUN_COMMAND_OUTPUT_MAX_BYTES,
    )
}

fn error_heuristic(text: &str) -> (bool, Option<String>) {
    for line in text.lines() {
        let lower = line.to_ascii_lowercase();
        let cli_error = lower.match_indices('%').any(|(index, _)| {
            let suffix = lower[index + 1..].trim_start();
            ["invalid input", "ambiguous command", "incomplete command"]
                .iter()
                .any(|prefix| suffix.starts_with(prefix))
        });
        if cli_error
            || lower.contains("syntax error")
            || lower.contains("unknown command")
            || line.trim() == "^"
        {
            return (true, Some(line.trim().to_owned()));
        }
    }
    (false, None)
}

fn finish_result(
    raw: &[u8],
    overflowed: bool,
    status: RunCommandStatus,
    exit_code: Option<i32>,
) -> RunCommandResult {
    let (output, truncated) = extract_output(raw, overflowed);
    let (likely_error, error_line) = error_heuristic(&output);
    RunCommandResult {
        output,
        exit_code,
        status,
        likely_error,
        error_line,
        truncated,
    }
}

pub async fn execute_run_command(
    params: &RunCommandParams,
    stream: &mut dyn TerminalCommandStream,
    nonce: &str,
) -> Result<RunCommandResult, AgentPortError> {
    if stream.initial_alternate_screen() {
        return Ok(finish_result(
            &[],
            false,
            RunCommandStatus::EnteredTui,
            None,
        ));
    }

    let use_posix = params.mode == RunCommandMode::Posix
        || (params.mode == RunCommandMode::Auto && exit_code_expression(params.shell).is_some());
    if params.mode == RunCommandMode::Cli {
        if let Some(command) = params.disable_paging_command.as_deref() {
            stream.send_input(&format!("{command}\r")).await?;
            discard_stream_updates_for(stream, Duration::from_millis(200)).await?;
        }
    }

    let prompt = match params.prompt.as_deref() {
        Some(pattern) => Some(
            Regex::new(pattern)
                .map_err(|error| AgentPortError::new(format!("invalid prompt regex: {error}")))?,
        ),
        None if params.mode == RunCommandMode::Cli => {
            prompt_pattern(last_non_empty_line(stream.initial_screen()))
        }
        None => None,
    };
    let expect = params
        .expect
        .as_deref()
        .map(Regex::new)
        .transpose()
        .map_err(|error| AgentPortError::new(format!("invalid expect regex: {error}")))?;

    if use_posix {
        let expression = exit_code_expression(params.shell).unwrap_or("$?");
        let marker = format!("printf '\\033]133;D;%s;tmex={nonce}\\033\\\\' \"{expression}\"");
        stream
            .send_input(&format!("{}; {marker}\r", params.command))
            .await?;
    } else {
        stream.send_input(&format!("{}\r", params.command)).await?;
    }

    let mut raw = Vec::new();
    let mut overflowed = false;
    let deadline = Instant::now() + params.timeout;
    let mut quiet_since = None;
    let mut previous_len = 0;

    loop {
        let now = Instant::now();
        if now >= deadline {
            break;
        }
        let wait = RUN_COMMAND_POLL_INTERVAL.min(deadline.saturating_duration_since(now));
        let update = stream.poll(wait).await?;
        let now = Instant::now();
        if update.alternate_screen {
            return Ok(finish_result(
                &raw,
                overflowed,
                RunCommandStatus::EnteredTui,
                None,
            ));
        }
        let remaining = RUN_COMMAND_OUTPUT_MAX_BYTES.saturating_sub(raw.len());
        if update.bytes.len() > remaining {
            overflowed = true;
        }
        raw.extend_from_slice(&update.bytes[..update.bytes.len().min(remaining)]);
        let cleaned = clean_terminal_text(&String::from_utf8_lossy(&raw));

        if expect
            .as_ref()
            .is_some_and(|pattern| pattern.find(&cleaned).is_some())
        {
            return Ok(finish_result(
                &raw,
                overflowed,
                RunCommandStatus::ExpectMatched,
                None,
            ));
        }
        if use_posix {
            if let Some(marker) = update.markers.iter().find(|marker| {
                marker.kind == 'D'
                    && marker
                        .params
                        .iter()
                        .any(|parameter| parameter == &format!("tmex={nonce}"))
            }) {
                return Ok(finish_result(
                    &raw,
                    overflowed,
                    RunCommandStatus::Completed,
                    marker.exit_code,
                ));
            }
        }

        let mut tail_start = cleaned.len().saturating_sub(200);
        while !cleaned.is_char_boundary(tail_start) {
            tail_start += 1;
        }
        let tail = &cleaned[tail_start..];
        if ["--More--", "<--- More --->", "More: <space>"]
            .iter()
            .any(|marker| tail.contains(marker))
            || tail.to_ascii_lowercase().contains("---(more")
        {
            stream.send_input(" ").await?;
            quiet_since = None;
            previous_len = raw.len();
            continue;
        }

        if let Some(prompt) = &prompt {
            let last_line = last_non_empty_line(&cleaned);
            let (output, _) = extract_output(&raw, overflowed);
            if !output.is_empty() && prompt.find(last_line).is_some() {
                return Ok(finish_result(
                    &raw,
                    overflowed,
                    RunCommandStatus::Completed,
                    None,
                ));
            }
        }

        if raw.len() == previous_len {
            let quiet_for = now.saturating_duration_since(*quiet_since.get_or_insert(now));
            let (output, _) = extract_output(&raw, overflowed);
            if quiet_for >= Duration::from_millis(600)
                && (!output.is_empty() || quiet_for >= Duration::from_millis(1500))
            {
                return Ok(finish_result(
                    &raw,
                    overflowed,
                    RunCommandStatus::Completed,
                    None,
                ));
            }
        } else {
            quiet_since = None;
            previous_len = raw.len();
        }
    }

    Ok(finish_result(
        &raw,
        overflowed,
        RunCommandStatus::Timeout,
        None,
    ))
}

#[cfg(test)]
mod tests {
    use super::*;

    struct ImmediateStream {
        sent_at: Vec<Instant>,
        settle_polls: usize,
        command_polls: usize,
    }

    #[async_trait]
    impl TerminalCommandStream for ImmediateStream {
        fn initial_screen(&self) -> &str {
            "$ "
        }

        fn initial_alternate_screen(&self) -> bool {
            false
        }

        async fn send_input(&mut self, _data: &str) -> Result<(), AgentPortError> {
            self.sent_at.push(Instant::now());
            Ok(())
        }

        async fn poll(&mut self, wait: Duration) -> Result<CommandStreamUpdate, AgentPortError> {
            if self.sent_at.len() == 1 {
                self.settle_polls += 1;
                if self.settle_polls > 3 {
                    tokio::time::sleep(wait).await;
                }
                return Ok(CommandStreamUpdate::default());
            }

            self.command_polls += 1;
            Ok(if self.command_polls == 40 {
                CommandStreamUpdate {
                    bytes: b"show\r\nok\r\n$ ".to_vec(),
                    ..CommandStreamUpdate::default()
                }
            } else {
                CommandStreamUpdate::default()
            })
        }

        async fn close(&mut self) {}
    }

    #[tokio::test]
    async fn immediately_ready_polls_do_not_consume_real_deadlines() {
        let mut stream = ImmediateStream {
            sent_at: Vec::new(),
            settle_polls: 0,
            command_polls: 0,
        };
        let mut params = RunCommandParams::new("show");
        params.mode = RunCommandMode::Cli;
        params.prompt = Some(r"\$\s*$".to_owned());
        params.disable_paging_command = Some("terminal length 0".to_owned());
        let result = execute_run_command(&params, &mut stream, "test-nonce")
            .await
            .expect("run command");

        assert_eq!(result.status, RunCommandStatus::Completed);
        assert!(result.output.contains("ok"));
        assert_eq!(stream.command_polls, 40);
        assert!(stream.sent_at[1].duration_since(stream.sent_at[0]) >= Duration::from_millis(200));
    }
}
